packages/@aws-cdk/lambda-layer-awscli/lib/awscli-layer.ts

```diff
@@ -1,19 +1,150 @@
+      if (result) {
+        logs.push(`require('${id}') succeeded.`);
+      }
+      return result;
+    } catch (e) {
+      logs.push(`require('${id}') failed.`);
+      const eAsError = e as Error;
+      if (eAsError?.stack) {
+        logs.push(eAsError.stack);
+      }
+    }
+  }
+
+  private static installAndLoadPackage(from: string, logs: string[]): any {
+    const installDir = AwsCliLayer.findInstallDir();
+    if (!installDir) {
+      logs.push('Unable to find an install directory. require.main.paths[0] is undefined.');
+      return;
+    }
+    logs.push(`Installing from: ${from} to: ${installDir}`);
+    childproc.execSync(`npm install ${from} --no-save --prefix ${installDir} -q`);
+    return AwsCliLayer.requireWrapper(path.join(installDir, 'node_modules', AwsCliLayer.assetPackageName, 'lib/index.js'), logs);
+  }
+
+  private static findInstallDir(): string | undefined {
+    if (!require.main?.paths) {
+      return undefined;
+    }
+    return require.main.paths[0];
+  }
+
   constructor(scope: Construct, id: string) {
-    super(scope, id, {
-      code: lambda.Code.fromAsset(path.join(__dirname, 'layer.zip'), {
+    const logs: string[] = [];
+    let fallback = false;
+
+    const targetVersion = AwsCliLayer.requireWrapper(path.join(__dirname, '../package.json'), logs).devDependencies[AwsCliLayer.assetPackageName];
+
+    let assetPackage = AwsCliLayer._tryLoadPackage(targetVersion, logs);
+
+    if (!assetPackage) {
+      const downloadPath = AwsCliLayer._downloadPackage(targetVersion, logs);
+      if (downloadPath) {
+        assetPackage = AwsCliLayer.installAndLoadPackage(downloadPath, logs);
+      }
+    }
+
+    let code;
+    if (assetPackage) {
+      // ask for feedback here
+      const asset = new assetPackage.AwsCliAsset(scope, `${id}-asset`);
+      code = lambda.Code.fromBucket(asset.bucket, asset.s3ObjectKey);
+    }
+
+    if (!code) {
+      fallback = true;
+      logs.push(`Unable to load ${AwsCliLayer.assetPackageName}. Falling back to use layer.zip bundled with aws-cdk-lib`);
+      code = lambda.Code.fromAsset(path.join(__dirname, 'layer.zip'), {
         // we hash the layer directory (it contains the tools versions and Dockerfile) because hashing the zip is non-deterministic
         assetHash: FileSystem.fingerprint(path.join(__dirname, '../layer')),
-      }),
+      });
+    }
+
+    super(scope, id, {
+      code: code,
       description: '/opt/awscli/aws',
     });
+
+    if (debugModeEnabled()) {
+      Annotations.of(this).addInfo(logs.join('\n'));
+    }
+
+    if (fallback) {
+      Annotations.of(this).addWarning(`WARNING! ACTION REQUIRED! Your CDK application is using ${this.constructor.name} and this construct may experience a breaking change for your environment in a future release. See https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.lambda_layer_awscli-readme.html or https://github.com/aws/aws-cdk/issues/999999999 for details and resolution instructions.`);
```

Comment:
Put the immediate suggested remediation in here instead of saying that "something" may break in the future. That sounds vague and scary.

Suggestion:

```
[ACTION REQUIRED] Your CDK application is using ${this.constructor.name}. Add a dependency on '@aws-cdk/asset-awscli-v1' @ ^2.0.0, or the equivalent in your language to remove this warning. See https://xxxx for more information.
```

One link, not two.

Bonus points if we can detect the jsii language and translate the package name, but that's not MVP.

Reply:
Got it, updated.

Reply:
Any thoughts on the wording for the notice? https://github.com/cdklabs/aws-cdk-notices/pull/21/files I don't think it's quite right yet.